File: 0143.ReorderList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        second = slow.next
        slow.next = None
        p = head
        q = self.reverse(second)
        while p and q:
            tp, tq = p.next, q.next
            p.next = q
            q.next = tp
            p, q = tp, tq

    def reverse(self, head):
        tail = None
        while head:
            temp = head
            head = head.next
            temp.next = tail
            tail = temp
        return tail

File: 0143.ReorderList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reorderList(head *ListNode)  {
    if head == nil || head.Next == nil {
        return
    }
    mid, tail := head, head.Next
    for tail != nil && tail.Next != nil {
        mid, tail = mid.Next, tail.Next.Next
    }
    head2 := mid.Next
    mid.Next = nil
    head2 = reverse(head2)
    merge(head, head2)
}

func reverse(head *ListNode) *ListNode {
    var newHead *ListNode
    for head != nil {
        nxt := head.Next
        head.Next = newHead
        newHead = head
        head = nxt
    }
    return newHead
}

func merge(h1, h2 *ListNode) {
    // len(h1) == len(h2) or len(h1) == len(h2)+1
    for h2 != nil {
        n1, n2 := h1.Next, h2.Next
        h1.Next = h2
        h2.Next = n1
        h1, h2 = n1, n2
    }
}